docs/upgrading_to_nameko_3.rst

```diff
@@ -0,0 +1,347 @@
+
+Accessing config via the global:
+
+.. code-block:: python
+
+    from nameko import config
+    from nameko.extensions import DependencyProvider
+
+    class Database(DependencyProvider):
+        def setup(self):
+            db_uris = config[DB_URIS_KEY]  # <- Nameko 3 way, won't work with Nameko 2!
+            # ...
+
+Accessing config via the service container:
+
+.. code-block:: python
+
+    from nameko.extensions import DependencyProvider
+
+    class Database(DependencyProvider):
+        def setup(self):
+            db_uris = self.container.config[DB_URIS_KEY]  # <- still works, Nameko 2 & 3
+            # ...
+
+In Nameko 3, the above will result in a `DeprecationWarning`:
+
+.. code-block:
+
+    .../nameko/containers.py:173: DeprecationWarning: Use ``nameko.config`` instead.
+    warnings.warn("Use ``nameko.config`` instead.", DeprecationWarning)
+
+
+Programmatically running services
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. warning:: Breaking Change!
+
+    The following section contains a breaking change.
+
+This section is relevant if you are creating runner programmatically, rather than using the `run` CLI command.
+
+In Nameko 3, the `ServiceContainer` and `ServiceRunner` classes do not accept the `config` argument anymore. You must update configuration in the global object, with `config.patch` or directly:
+
+.. code-block:: python
+
+    import nameko
+    from nameko.containers import ServiceContainer
+
+    nameko.config["AMQP_URI"] = "pyamqp://someuser:*****@somehost/"
+
+    container = ServiceContainer(Service)
+    container.start()
+    # AMQP extensions will connect as someuser to RabbitMQ broker at somehost ...
+    container.stop()
+
+Same for the runner:
+
+.. code-block:: python
+
+    from nameko.runners import ServiceRunner
+
+    nameko.config["AMQP_URI"] = "pyamqp://someuser:*****@somehost/"
+
+    runner = ServiceRunner()
+    runner.add_service(ServiceA)
+    runner.add_service(ServiceB)
+
+    runner.start()
+    # AMQP extensions will connect as someuser to RabbitMQ broker at somehost ...
+    runner.stop()
+
+Other configuration related changes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* The `--broker`` CLI option is deprecated in favour of `--define` and `--config`. TODO: doesn’t actually raise a warning anymore
+* The built-in Config dependency provider is deprecated as the config can be accessed and read directly.
+
+Testing services
+----------------
+
+In Nameko 2, configuration was always explicitly passed as an object, which made it easy to manage in tests. The global config object needs some different ways of managing configuration in tests.
+
+Config patching
+~~~~~~~~~~~~~~~
+
+The global `config` object has a `patch` helper object which can be used as a context manager or as a decorator. Here is an example of patching the config object in a pytest fixture:
```

Comment:
As the config global object is basically just a dictionary, there is nothing to prevent the user from simply changing its values with a `config["blah"] = 123`. Please add an explanation as to why `patch()` is important. Namely because it only applies the patched changes to the current execution context and will revert to the initial values once out of the context.